### ISSUE TITLE
"systemd-run --wait" exit code if process killed by a signal

### ISSUE THREAD

**Issue by rdiez** [systemd/systemd#22812]
I have the following script called kill-myself-with-sigterm.sh :

```
#!/bin/bash

kill -SIGTERM "$$"
```

Then I ran it like this:

```
$ systemd-run --user --wait ./kill-myself-with-sigterm.sh
Running as unit: run-u721.service
Finished with result: success
Main processes terminated with: code=killed/status=TERM
```

I was surprised that the exit code from systemd-run was nevertheless 0, indicating success.

I tried to find the reason in the documentation, but I couldn't understand it. This is an area that, in my opinion, needs to be expressed in a way that is easier to understand for non-expert users like me.

I did some more investigation. If the script terminates with exit code 123, then systemd-run yields that same exit code, as documented. It's only the case "killed by a signal" that yields 0. The actual signal does not seem to matter, for SIGHUP behaves the same.

In my opinion, that behaviour is unexpected. "systemd-run --wait" should yield a non-zero exit code if the transient unit dies because of a signal. Otherwise, it is not easy to determine whether the child process did succeed or just got killed.

Apparently, I am not the only one to scratch his head:

[How to tell if a systemd-run service was killed/died versus exiting normally?](https://unix.stackexchange.com/questions/543429/how-to-tell-if-a-systemd-run-service-was-killed-died-versus-exiting-normally)

Then I changed the service type to 'exec' like this:

`systemd-run --user --wait --service-type=exec  ./kill-myself-with-sigterm.sh`

The behaviour is apparently the same. But with 'oneshot', the behaviour is different:

`systemd-run --user --wait --service-type=oneshot  ./kill-myself-with-sigterm.sh`

If the transient unit terminates with an exit code of 0, systemd-run yields 0 too.

But if the transient unit terminates with a non-zero exit code, then systemd-run prints the following message and always yields 1, instead of the actual status code:

`Job for run-u728.service failed because the control process exited with error code.`

If the transient unit is killed by a signal, then systemd-run prints the following message and always yields a status code of 1:

`Job for run-u727.service failed because a fatal signal was delivered to the control process.`

This behaviour does help in my particular usage scenario, but the original status code is lost. That status code could be important. For example, 'grep' can yield 0, 1 or 2, and 1 is not an error.

So there is apparently no wait to get the exact status code in case of normal termination, and some non-zero status code in case of dying from a signal, like it is usually the case with Bash etc.

In any case, these differences in behaviour could be documented in simpler terms, in my opinion, at least in the case of systemd-run.

I am using systemd 245 (245.4-4ubuntu3.15) that comes with Ubuntu 20.04.4, but I think that this issue is actually rather independent from the systemd version.

**Comment by poettering** [comment]
systemd-run will return 0 on successful exit as defined by the `SuccessExitStatus=` setting, as configurable for any service. See systemd.service(5) man page for details.

On non-successful exit (by default any non-zero exit code is that) will propagate the original exit code. On failure by by signal will return 255, any other failure wil return 1.

And this is more or less described in the man page systemd-run. Quoting:

> On success, 0 is returned. If systemd-run failed to start the service, a non-zero return value will be returned. If systemd-run waits for the service to terminate, the return value will be propagated from the service. 0 will be returned on success, including all the cases where systemd considers a service to have exited cleanly, see the discussion of SuccessExitStatus= in systemd.service(5).

**Comment by rdiez** [comment]
First of all, thanks for your answer.

I think I understand now. There is no way for systemd-run to do these 2 at the same time:
a) propagate the original exit code
b) consider SIGHUP, SIGINT, SIGTERM and SIGPIPE as failure (non-zero exit code).

There are only 2 options:

Option 1) "systemd-run --user --wait" will create by default a transient unit of type 'simple', and then SIGHUP, SIGINT, SIGTERM and SIGPIPE will be considered as successful (exit code 0), as documented for "SuccessExitStatus=".

That is no good because the user does not easily know that one of those signals killed the process (which is normally an indication of failure by death, and not success).

The bit "any other failure will return 1" is also problematic, because tools like 'grep' tend to assign meaning to the lower exit code values.

Option 2) "systemd-run --user --wait --service-type=oneshot" will consider death by any signal as failure (exit code 1), but, if the child terminates normally, the original exit code will not be propagated.

The exit code from systemd-run with 'oneshot' will always be either:
- Exit code 0 (child exited with status 0), or
- Exit code 1 (child exited with status != 0, killed by signal, or any other failure).

That is no good either because the process' exit code may be meaningful, like in the case of 'grep'.

Is that correct?

By the way, you said "failure by signal will return 255". Shouldn't that be properly documented? I could not find it anywhere. The closest I could find on this page:

[https://www.freedesktop.org/software/systemd/man/systemd.service.html](https://www.freedesktop.org/software/systemd/man/systemd.service.html)

is

`However, if an ExecCondition= command exits with 255 or abnormally (e.g. timeout, killed by a signal, etc.), the unit will be considered failed`

But that does not really say that "killed by a signal" will yield an exit code of 255, at least when using systemd-run.
